### PR TITLE
[Kolors] Add PAG

### PR DESCRIPTION
**Pull request by asomoza**
# What does this PR do?

Adds PAG to kolors

Note: Kolors with PAG behaves different than with SDXL, I had to use a lower scale and different layers to get good results.

### Txt2Img
```python
import torch

from diffusers import AutoPipelineForText2Image, DPMSolverMultistepScheduler


pipe = AutoPipelineForText2Image.from_pretrained(
    "Kwai-Kolors/Kolors-diffusers",
    torch_dtype=torch.float16,
    variant="fp16",
    enable_pag=True,
    pag_applied_layers=["down.blocks.2.attentions.1", "up.blocks.0.attentions.1"],
).to("cuda")
pipe.scheduler = DPMSolverMultistepScheduler.from_config(pipe.scheduler.config, use_karras_sigmas=True)

prompt = "cinematic film still of a pirate ship is navigating the milk inside of a steaming cup of coffee. The ship’s wooden hull glistens with droplets of coffee, and tiny pirates can be seen on deck, adjusting sails and peering through miniature telescopes. The tilt-shift effect adds a magical touch, high budget hollywood movie, cinemascope, epic, film grain"

image = pipe(
    prompt=prompt,
    negative_prompt="",
    guidance_scale=5.5,
    num_inference_steps=25,
    pag_scale=1.5,
).images[0]
 ```

|W/O|PAG|
|---|---|
|||

After the PAG refactor, the image changed a bit, not sure why:

|before|after|
|---|---|
|||

## Who can review?

@yiyixuxu @a-r-r-o-w @sunovivid @sayakpaul

### CODE REVIEW

File: src/diffusers/pipelines/pag/pipeline_pag_kolors.py [hunk 1]
```diff
@@ -0,0 +1,1130 @@
...
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers import KolorsPipeline
+
+        >>> pipe = KolorsPipeline.from_pretrained(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pipeline_pag_kolors.py:52]: update this?

File: src/diffusers/pipelines/pag/pipeline_pag_kolors.py [hunk 2]
```diff
@@ -0,0 +1,1130 @@
...
+        self.vae_scale_factor = (
+            2 ** (len(self.vae.config.block_out_channels) - 1) if hasattr(self, "vae") and self.vae is not None else 8
+        )
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+
+        self.default_sample_size = self.unet.config.sample_size
+
+        self.set_pag_applied_layers(pag_applied_layers)
+
+    def encode_prompt(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pipeline_pag_kolors.py:210]: copied from?

File: src/diffusers/pipelines/pag/pipeline_pag_kolors.py [hunk 3]
```diff
@@ -0,0 +1,1130 @@
...
+        if accepts_eta:
+            extra_step_kwargs["eta"] = eta
+
+        # check if the scheduler accepts generator
+        accepts_generator = "generator" in set(inspect.signature(self.scheduler.step).parameters.keys())
+        if accepts_generator:
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
+
+    def check_inputs(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pipeline_pag_kolors.py:463]: copied from?

**Review by yiyixuxu — APPROVED** [review]
thanks!
add a test?

**Comment by asomoza** [comment]
Added the fast tests and I fixed the old ones because I needed to get the original Kolors Pipeline to test the AutoPipeline. In short, I had to add all the missing tests I had as TODO for kolors.

@yiyixuxu @sayakpaul should I add the slow tests for PAG? this one has a really big text encoder which is not in transformers yet.

**Comment by sayakpaul** [comment]
> @yiyixuxu @sayakpaul should I add the slow tests for PAG? this one has a really big text encoder which is not in transformers yet.

I would say, it'd be okay skipping the SLOW tests for now.

**Comment by a-r-r-o-w** [comment]
We should try and merge this before next release, no?

File: src/diffusers/pipelines/kolors/tokenizer.py [hunk 4]
```diff
@@ -143,10 +143,18 @@ def get_command(self, token):
     def unk_token(self) -> str:
         return "<unk>"
 
+    @unk_token.setter
+    def unk_token(self, value: str):
+        self._unk_token = value
```
Comment (sayakpaul) [inline src/diffusers/pipelines/kolors/tokenizer.py:148]: Seems like these should have been done in a separate PR?

File: tests/pipelines/kolors/test_kolors.py [hunk 5]
```diff
@@ -133,23 +133,11 @@ def test_inference(self):
         max_diff = np.abs(image_slice.flatten() - expected_slice).max()
         self.assertLessEqual(max_diff, 1e-3)
 
-    # throws AttributeError: property 'eos_token' of 'ChatGLMTokenizer' object has no setter
-    # not sure if it is worth to fix it before integrating it to transformers
     def test_save_load_optional_components(self):
-        # TODO (Alvaro) need to fix later
-        pass
+        super().test_save_load_optional_components(expected_max_difference=2e-4)
```
Comment (sayakpaul) [inline tests/pipelines/kolors/test_kolors.py:137]: Same. Probably should have been done in a separate PR.

**Review by sayakpaul — APPROVED** [review]
LGTM. Let's add the pipeline to the list of pipelines we support in the [docs](https://huggingface.co/docs/diffusers/main/en/api/pipelines/pag)?

File: src/diffusers/pipelines/kolors/tokenizer.py (hunk 4 above)
Comment (asomoza) [reply src/diffusers/pipelines/kolors/tokenizer.py:148]: if I don't do this, the tests fails (throws AttributeError), I'm ok with doing this in a separate PR but then I'll have to skip the tests in this PR

Comment (sayakpaul) [reply src/diffusers/pipelines/kolors/tokenizer.py:148]: Ah okay. Then no worries.
